GSDX: Improve CRTC width detection

ssakash opened the pull request:
**Summary of changes:**

* Prevent usage of non-standard / weird DW register values.

Fixes #1073 

**Master branch:**

![broken](http://ultraimg.com/images/TimeCrisis_Broken.png)

**Pull request branch:**

![fixed](http://ultraimg.com/images/TimeCrisis_Fixed.png)

**Build download link:** [here](https://ci.appveyor.com/api/buildjobs/x1wt04c52m4249an/artifacts/pcsx2-v1.5.0-dev-444-g39a0e19-master-652-vs2015-Win32-AppVeyor.7z)

ssakash commented:
@FlatOutPS2

Could you do some regression checking ? look out for games with a very low width value ;)

**Additional request:**

could you try removing the [following comment](https://github.com/PCSX2-Alpha/pcsx2/blob/Time_Crisis/plugins/GSdx/GSState.cpp#L381) and go through some games and check whether any games print a weird DW value (lower than 2559) on the console log ? Thanks !

refractionpcsx2 commented:
Need to check vertical shooters with this, make sure they aren't broken as they use a very thin width (i'm presuming, I've not actually checked)

willkuer commented:
Reading your comment i have the feeling you completely understand all issues and know input as well as expected output. Why does it not work though? Why do you need a hack?

gregory38 commented on plugins/GSdx/GSState.cpp:
```diff
@@ -349,6 +349,19 @@ void GSState::ResetHandlers()
+	// games which actually rely on 320 for height always use the default DW value with a magnification factor of 7 instead.
+	if (m_regs->DISP[i].DISPLAY.DW < 1300 && m_regs->DISP[i].DISPLAY.MAGH >= 3 && !Vmode_VESA_DTV)
+	{
+		#if _DEBUG
```
GL_INS is only enabled in debug/dev build so no need of ifdef.

gregory38 commented:
Hum, could you post both the standard and non stardard dw/dh value with their associated magnification factor.

gregory38 commented on plugins/GSdx/GSState.cpp:
```diff
@@ -349,6 +349,19 @@ void GSState::ResetHandlers()
+	// Time crisis uses a very weird non-standard DW register value at few scenes like the splitscreen mode.
+	// This makes the game only show half of it's display rectangle eventually blocking the other part of screen.
+	// the wrong width (320) is produced by using a very low DW value with a magnification factor of 3.
+	// games which actually rely on 320 for height always use the default DW value with a magnification factor of 7 instead.
```
by "height" do you mean width?

FlatOutPS2 commented:
I've tried several games; Baldur's Gate DA demo, R: Racing, Everbody's Golf, ICO, Disgaea, Ridge Racer V, Fatal Fury BAV1, Tourist Trophy, including available splitscreen modes and all had 2559 and showed no regressions.

Resident Evil 4 was a little different, which had 2561 on 50hz, 2559 on 60hz and 1279 halfsies on progressive. But it was fine too.

gregory38 commented:
Add a CRC, because so far it is voodoo stuff.

ssakash commented:
Nope, no **voodoo stuffs** involved.  (what the heck is ``voodoo`` anyway ? )

![doc](http://ultraimg.com/images/Screenshot2023c49.png)

> ----------------.-----------.-------.-------.-------.-------.-------.
 Mode           | DX        | DY    | MAGH  | MAGV  | DW    | DH    |
----------------+-----------+-------+-------+-------+-------+-------|
  0: NTSC-NI    | 632       | 25    | 3     | 0     | 2559  | 239   |
  1: NTSC-I     | 632       | 50    | 3     | 0     | 2559  | 479   |
  2: PAL-NI     | 652       | 36    | 3     | 0     | 2559  | 287   |
  3: PAL-I      | 652       | 72    | 3     | 0     | 2559  | 575   |
  4: VESA-1A    | 276 (272)¹| 34    | 1     | 0     | 1279  | 479   |
  5: VESA-1C    | 356 (352) | 18    | 1     | 0     | 1279  | 479   |
  6: VESA-2B    | 420 (416) | 26    | 1     | 0     | 1599  | 599   |
  7: VESA-2D    | 468 (464) | 23    | 1     | 0     | 1599  | 599   |
  8: VESA-3B    | 580 (576) | 34    | 1     | 0     | 2047  | 767   |
  9: VESA-3D    | 260 (256) | 30    | 0     | 0     | 1023  | 767   |
 10: VESA-4A    | 348 (344) | 40    | 0     | 0     | 1279  | 1023  |
 11: VESA-4B    | 348 (376) | 40    | 0     | 0     | 1279  | 1023  |
 12: DTV-480P   | 232 (228) | 35    | 1     | 0     | 1439  | 479   |
 13: DTV-1080I  | 238 (234) | 40    | 0     | 0     | 1919  | 1079  |
 14: DTV-720P   | 302 (298) | 24    | 0     | 0     | 1279  | 719   |
----------------^-----------^-------^-------^-------^-------^-------'

Thanks for the test @FlatOutPS2 ! you confirmed my assumption. It's fine for progressive modes to use 1279 as we use different video modes on progressive mode (VESA,DTV). please look at the above table for reference. at progressive mode we don't enter the half-width case condition due to the ``!Vmode_VESA_DTV`` check.

> Reading your comment i have the feeling you completely understand all issues and know input as well as expected output. Why does it not work though? Why do you need a hack?

I'm not sure about the source of issue. I couldn't tell you much without doing some hardware tests on the PS2. but we do know that the game emitting half of the standard DW is wrong behavior on NTSC/PAL video modes.

> Add a CRC, because so far it is voodoo stuff.

I personally prefer having it at general code. (it should to be safe)

ssakash commented:
Oh, also I won't probably be available/active for quite a while due to health issues. so please be patient if I don't quickly reply to your messages, I'll get to them eventually ;) 

> Need to check vertical shooters with this, make sure they aren't broken as they use a very thin width (i'm presuming, I've not actually checked)

Hmm, can you do some checks on them ? would be interesting to find out some more unique cases.

refractionpcsx2 commented:
I don't have any xD

FlatOutPS2 commented:
I believe 90% of all PS2 vertical shooter were exclusive to Japan. Don't check Gamefaqs to confirm, it lists most of the horizontal shooters as vertical. :/

refractionpcsx2 commented:
Psyvariar Revision
Espgaluda
Homura
Strikers 1945/Gunbird
Raiden III

Thats some of the ones I'd be worried about, where they purposefully have black borders either side. Unfortunately I own none of these.

gregory38 commented:
voodoo is a Religion (relying on black magic) of some countries.

Why do you said the value of DW is incorrect? Your tabs above are only example.

```
Psyvariar Revision
```
I vaguely remember this name so potentially I have it.

ssakash commented:
> Why do you said the value of DW is incorrect? Your tabs above are only example.

Well, personally I think it's incorrect due to the following two reasons :

* It only shows the half of display rectangle to be displayed for the following game according to current wrong DW value.
* All other observed games so far seem to use hardcoded DW value on specific video modes `` NTSC/PAL: (2559) , VESA: (1279) , 480P: (1439) , 1080I: (1919) ``, the only other different value observed so far on NTSC/PAL video modes is ``(2561)``

Hence, It's highly likely that the DW value is _wrong_ for the following game. though maybe it's plausible that some insane dev might consider using a different value to get more possibilities with scaling the resolution using magnification factors (MAGH , MAGV).

Either way, someone should really look into these vertical shooter games on whether the patch fixes/causes any issues.

> Psyvariar Revision
Espgaluda
Homura
Strikers 1945/Gunbird
Raiden III

unfortunately, I haven't even heard any one of these games !

refractionpcsx2 commented:
What is the reason the value is so small is because the VCK fequency is low (30fps?)? as DW is a measure of VCK cycles, not pixels.

ssakash commented:
> What is the reason the value is so small is because the VCK fequency is low (30fps?)? as DW is a measure of VCK cycles, not pixels.

I'm not sure something like that is possible, VCK Frequency is harcoded and set by the kernel dependent on video mode. you can't dynamically infer with it without switching through different video modes which the game doesn't seem to be doing.

refractionpcsx2 commented:
no i mean are you just interpreting it wrong, instead of saying "that value is obviously wrong, lets hack it to something more sensible" it's very unlikely any value is wrong, but probably dependant on how the value is interpreted.

The manual says the DW is a measure of VCLK, maybe the value is lower, but the VCLK value for the video mode it is in could also be lower, which would make the picture wider (or at least what it normally is), because at the moment, what is there is essentially a hack, but there has to be a reason why that value is lower than normal.

ssakash commented:
> t could be that it is because the 2 screens are drawn separately and we can only see the one because of how gsdx currently handles it, I mean this is a specific case for "split screen" gameplay.

@refractionpcsx2  seems to have nailed the issue :) , redone my patch with respect to refraction's suggestion and it seems to be the case with time crisis 2 and does fix the issue !

@FlatOutPS2 : your usual regression checking if you don't mind ;)

refractionpcsx2 commented:
Still got that fixed value in there, is there not another way to check if that value is for half the screen?

but good job :)

turtleli commented:
Is the special case needed? I think it is GSRenderer::Merge() that is unable to deal with two side by side rectangles. If I understand the code correctly, this:
```
		dst[i] = GSVector4(off).xyxy() + scale * GSVector4(r.rsize());

		fs.x = max(fs.x, (int)(dst[i].z + 0.5f));
		fs.y = max(fs.y, (int)(dst[i].w + 0.5f));
```
is the reason for half the image being cut off.

ssakash commented:
> Is the special case needed? I think it is GSRenderer::Merge() that is unable to deal with two side by side rectangles. If I understand the code correctly, this:
        ``dst[i] = GSVector4(off).xyxy() + scale * GSVector4(r.rsize());``
        ``fs.x = max(fs.x, (int)(dst[i].z + 0.5f));``
       ``fs.y = max(fs.y, (int)(dst[i].w + 0.5f));``
is the reason for half the image being cut off.

``Merge()`` is wrong since we don't consider the total width for the the two rectangles on ``GetDisplayRect()`` , the correct fix should be on ``GetDisplayRect()` which is the source of issue.

> Still got that fixed value in there, is there not another way to check if that value is for half the screen?

That check isn't really needed now but doesn't hurt to be on the safe side ;) also it's useful for identifying such similar cases on other games. (debug messages)

turtleli commented:
> Merge() is wrong since we don't consider the total width for the the two rectangles on GetDisplayRect() , the correct fix should be on `GetDisplayRect() which is the source of issue.

Instead of using a special case, I was thinking something like this might work for the width and height calculations when both outputs are enabled:
```
width  = max(dx[0] + dw[0], dx[1] + dw[1]) - min(dx[0], dx[1])
height = max(dy[0] + dh[0], dy[1] + dh[1]) - min(dy[0], dy[1])
```

ssakash commented:
> width  = max(dx[0] + dw[0], dx[1] + dw[1]) - min(dx[0], dx[1])

This should work for the split screen mode but I'm afraid it'll break the width handling at cases where the second rect is unused.

_**example:**_ at cases where, dw[0] = 0; dw[1] = 640; dx[0]=0; dx[1]=150; (Including the magnification scaling) , you'll calculate the width to 790 which is the wrong value but I do understand your idea though ;)

oh also, I assume you wanted to include the magnification factors with your equation ? without them your code would be wrong by the way.

turtleli commented:
The code was an example to highlight the main idea, adding magnification/single rect stuff would make the idea less clear.

ssakash commented:
I've updated the code to only detect the usage of two rectangles and double the width when necessary, removed the fixed DW value. also used some macros to make the code more readable.

**Note:** Code isn't placed on ``merge()`` since it'll give you wrong width value at the GS Frame, so made them on ``GetDisplayRect()``. Possibly it'd be nice to have the code generally as @turtleli stated but I'm afraid of the possibility of regressions on games which do strange things with the rects. so currently the code only handles the cases where two identical rects are used for rendering, won't work if 70% of rendering is done by one rect and 30% by another one. ( I'm not sure if any games even do something like that, so unless we discover such cases I personally prefer only handling the cases of identical rects)

turtleli commented:
Could you provide a GS dump? It'd probably be quite useful to have.

ssakash commented:
> Could you provide a GS dump?

[Here you go](https://www.dropbox.com/s/x86c8lxptzk8d7f/Timecrisis_dumps.7z?dl=0)

gregory38 commented:
I think a comment that say ".. use 2 display rectangles `side by side`.. So code will merge both screen into a single display...".

So if I understand correctly, it also mean that game will push 2 half rectangle to the screen. It could explain the not-standard config

gregory38 commented:
As a side note, maybe we can/must also check that both output are enabled ?

ssakash commented:
I'll push a much cleaner fix after the laptop which I had ordered arrives. Probably at the first week of May..(Yes, I  had recently moved out to a new home and I'm away from my desktop :/ )

ssakash commented:
@gregory38 

We already check whether DW value of both rectangles are similar so I'm not sure whether it's necessary to check if both outputs are enabled.

gregory38 commented:
But what will happen if a game sets some values in DW but doesn't enable the output.

ssakash commented:
Oh. I didn't consider that possibility.. I haven't seen any games do
anything like that. I guess I'll also do a check for the status of output
circuit to be on the safe side.

Sorry though, I can only update the PR after my laptop arrives :/ ( which
would probably be around next month. ..)
On 24 Apr 2016 15:28, "Gregory Hainaut" <notifications@github.com> wrote:

But what will happen if a game sets some values in DW but doesn't enable
the output.

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub
<https://github.com/PCSX2/pcsx2/pull/1234#issuecomment-213922683>

ssakash commented:
**Note**: I have an idea to have the fix on ``merge()`` and also keep the scaling code / GS Frame title bar resolution right. The idea is to scale the code based on the merged output rect rather than the frame memory. I'll push the fix at the specific time mentioned earlier.

turtleli commented:
Well, the merge part would be something like turtleli@4ca31387de78c644c3196ae16f9935f44c48ebca , I was too lazy to adjust GetDisplayRect in the same way (GS frame title + SW dumps require it).

ssakash commented:
Upscaling code is also dependent on ``GetDisplayRect()``. however there are lots of bugs present on scaling code for custom resolution ( also need to fix that ). Thanks for the commit ! (Looks good to me ) , I'll cherry pick it along  with my follow up fixes.

ssakash commented:
I updated the code, please do some reviews and testing.

FlatOutPS2 commented:
No regressions so far, I'll try a few more games I also tried last time just to be sure.

ramapcsx2 commented:
I have the game Espgaluda, which is a vertical "bullet hell" shooter.
It allows to rotate the output to replicate the arcade experience.
Anyway, there aren't any issues with it on master, except for some nasty scaling artifacts.

gregory38 commented:
Could I merge it?

ssakash commented:
Well, I wanted to add some other PCRTC stuffs but I guess it'd be better to have it on a new pull request. (I don't want to trouble the testers with another round of regression checking :P )

Feel free to merge this one.

refractionpcsx2 commented:
This PR caused issues in FFXII with an offset overlay and bugs around the edges.

PIcture of it here: https://scontent-lhr3-1.xx.fbcdn.net/v/t1.0-0/s480x480/13394154_815152975256957_4235498632075698690_n.jpg?oh=a87c074e2b7cf46e42ea25fc97a52280&oe=57CC2A0E

Apparently happens in native and upscaled resolutions.  

@ssakash you can see more info on it on the Facebook group, posted by Jacob Stanbery :)

gregory38 commented:
@ssakash 
Could you open a new issue with regression flag and link it to this PR. Thanks you.

FlatOutPS2 commented:
It's the line `off.x = tex[i]->GetScale().x * (dr[i].left - baseline.x);` in GSRender.cpp that's the culprit. Unlike the y alternative there's no if statement before it, it might need one. :p

ssakash commented:
> Could you open a new issue with regression flag and link it to this PR. Thanks you.

I just got a GS dump of this issue, I'll make a new issue and push a fix tomorrow ( unless, **turteli** fixes it before me). way past my bed time right now.

turtleli commented:
Well, I don't have a GS dump for the issue, so I doubt I'll fix it before you. :P

Would be nice to have some kind of GS dump autotest/diff stuff I suppose.

ssakash commented:
> Well, I don't have a GS dump for the issue, so I doubt I'll fix it before you. :P

Heh, I wouldn't want an unfair challenge ;)

**GSDump**: [FFXII](https://www.dropbox.com/s/dhn7zam59swu0cn/FFXII.rar?dl=0)

ssakash commented:
I still haven't looked into it but would probably take care of it after finishing my lunch.

MrCK1 commented:
Come someone remind me what the number displayed at the top of the GS window represents?

Some of my games return values that appear to be offset by 1 pixel?

For instance, 

- Cartoon Network Racing reports `640x447`

- Wipeout Fusion is running at `640x447` (only in the menus) 

- Guitar Hero 2 is `512x447` and `640x447` in Prog Scan

Also, in a [comparison between GT4 and GT5](http://www.eurogamer.net/articles/digitalfoundry-gran-turismo-5-tech-analysis); the article stated that GT4's native resolution is `632x448` but in PCSX2 it is `640x448`

refractionpcsx2 commented:
> Also, in a comparison between GT4 and GT5; the article stated that GT4's native resolution is 632x448 but in PCSX2 it is 640x448

that's probably a rounding thing on our part, but they are generally the native resolutions, 640x447 (should be 448) are common for NTSC
